src/main/java/com/google/devtools/build/lib/analysis/PlatformSemantics.java

```diff
@@ -32,4 +40,27 @@
                 .nonconfigurable("Used in toolchain resolution")
                 .value(ImmutableList.of()));
   }
+
+  /**
+   * Return the target-specific execution platform constraints, based on the rule definition and any
+   * constraints added by the target.
+   */
+  public static ImmutableSet<Label> getExecutionPlatformConstraints(Rule rule) {
+    NonconfigurableAttributeMapper mapper = NonconfigurableAttributeMapper.of(rule);
+    Set<Label> execConstraintLabels = new HashSet<>();
```

Comment:
Is the nondeterministic HashSet okay in this context?

What about just using an ImmutableSet.Builder for the whole thing?

Reply:
Changed to ImmutableSet.Builder.